mithril-common/src/crypto_helper/merkle_map.rs

```diff
@@ -181,20 +187,35 @@ impl<K: MKMapKey, V: MKMapValue<K>> MKMap<K, V> {
         }
 
         let leaves_by_keys = self.group_leaves_by_keys(leaves);
-        let mut sub_proofs = BTreeMap::<K, MKMapProof<K>>::default();
-        for (key, sub_leaves) in leaves_by_keys {
-            if let Some(value) = self.get(&key) {
-                if let Some(proof) = value.compute_proof(&sub_leaves)? {
-                    sub_proofs.insert(key.to_owned(), proof);
+        let sub_proofs = leaves_by_keys
+            .into_par_iter()
+            .map(|(key, sub_leaves)| {
+                if let Some(value) = self.get(&key) {
+                    if let Some(proof) = value.compute_proof(&sub_leaves)? {
```

Comment:
It seems to be possible to simplify that by returning an Option<(key, value) instead of (key, Option<value>)
Then, the `fold` can be replace by a flatten.collect
I let you see if it's valid
   ```
        let sub_proofs: BTreeMap<_, _> = leaves_by_keys
            .into_par_iter()
            .map(|(key, sub_leaves)| {
                if let Some(value) = self.get(&key) {
                    if let Some(proof) = value.compute_proof(&sub_leaves)? {
                        return Ok(Some((key, proof)));
                    }
                }
                Ok(None)
            })
            .collect::<StdResult<Vec<_>>>()?
            .into_iter()
            .flatten()
            .collect();
```

Reply:
Excellent idea :+1: It works and it's much easier to read.